File: bcn_myanmar_hr_compliance/public/js/hrms/salary_structure.js
// Copyright (c) 2025, Business Centric Network Company Limited and Contributors
// License: MIT. See license.txt

frappe.ui.form.on("Salary Structure", {
	onload: function(frm) {
		frm.events.set_query_contributions_component(frm);
	},
	refresh: function (frm) {
		frm.events.set_query_contributions_component(frm);
	},
	
	company: function (frm) {
		frm.events.set_query_contributions_component(frm);
	},

	set_query_contributions_component: function (frm) {
		if (!frm.doc.company) return;		
		frm.set_query("salary_component", "custom_bcn_contributions", function () {
			return {
				filters: { component_type: "contribution", company: frm.doc.company },
				query: "hrms.payroll.doctype.salary_structure.salary_structure.get_salary_component",
			};
		});
	},

	calculate_totals: function (frm) {		
		var tblcontri = frm.doc.custom_bcn_contributions || [];
		
		var total_contributions = 0;
		for (var i = 0; i < tblcontri.length; i++) {
			total_contributions += flt(tblcontri[i].amount);
		}
		
		frm.doc.custom_bcn_total_contribution = total_contributions;
		frm.refresh_field("custom_bcn_total_contribution")
	}
});

frappe.ui.form.on("BCN Contribution Detail", {
	form_render: function (frm, cdt, cdn) {
		const row = locals[cdt][cdn];
		hrms.payroll_utils.set_autocompletions_for_condition_and_formula(frm, row);
	},

	amount: function (frm) {
		frm.events.calculate_totals(frm);
	},

	custom_bcn_contribution_remove: function (frm) {
		frm.events.calculate_totals(frm);
	},

	formula: function (frm, cdt, cdn) {
		const row = locals[cdt][cdn];
		if (row.formula && !row?.amount_based_on_formula && !frm.alerted_rows.includes(cdn)) {
			frappe.msgprint({
				message: __(
					"{0} Row #{1}: {2} needs to be enabled for the formula to be considered.",
					[toTitle(row.parentfield), row.idx, __("Amount based on formula").bold()],
				),
				title: __("Warning"),
				indicator: "orange",
			});
			frm.alerted_rows.push(cdn);
		}
	},

	salary_component: function (frm, cdt, cdn) {
		var child = locals[cdt][cdn];
		if (child.salary_component) {
			frappe.call({
				method: "frappe.client.get",
				args: {
					doctype: "Salary Component",
					name: child.salary_component,
				},
				callback: function (data) {
					if (data.message) {
						var result = data.message;
						frappe.model.set_value(cdt, cdn, "condition", result.condition);
						frappe.model.set_value(
							cdt,
							cdn,
							"amount_based_on_formula",
							result.amount_based_on_formula,
						);
						if (result.amount_based_on_formula == 1) {
							frappe.model.set_value(cdt, cdn, "formula", result.formula);
						} else {
							frappe.model.set_value(cdt, cdn, "amount", result.amount);
						}
						frappe.model.set_value(
							cdt,
							cdn,
							"statistical_component",
							result.statistical_component,
						);
						frappe.model.set_value(
							cdt,
							cdn,
							"depends_on_payment_days",
							result.depends_on_payment_days,
						);
						frappe.model.set_value(
							cdt,
							cdn,
							"do_not_include_in_total",
							result.do_not_include_in_total,
						);
						frappe.model.set_value(
							cdt,
							cdn,
							"variable_based_on_taxable_salary",
							result.variable_based_on_taxable_salary,
						);
						frappe.model.set_value(
							cdt,
							cdn,
							"is_tax_applicable",
							result.is_tax_applicable,
						);
						frappe.model.set_value(
							cdt,
							cdn,
							"is_flexible_benefit",
							result.is_flexible_benefit,
						);						
						frappe.model.set_value(
							cdt,
							cdn,
							"is_one_time_contribution",
							result.custom_bcn_is_one_time_contribution,
						);						
						refresh_field("custom_bcn_contributions");
					}
				},
			});
		}
	},

	amount_based_on_formula: function (frm, cdt, cdn) {
		var child = locals[cdt][cdn];
		if (child.amount_based_on_formula == 1) {
			frappe.model.set_value(cdt, cdn, "amount", null);
			const index = frm.alerted_rows.indexOf(cdn);
			if (index > -1) frm.alerted_rows.splice(index, 1);
		} else {
			frappe.model.set_value(cdt, cdn, "formula", null);
		}
	},
});

File: bcn_myanmar_hr_compliance/bcn_myanmar_payroll/overrides/salary_structure.py
# Copyright (c) 2025, Business Centric Network Company Limited and Contributors
# License: MIT. See license.txt


import re

import frappe
from frappe import _
from frappe.utils import cstr
from hrms.payroll.doctype.salary_structure.salary_structure import SalaryStructure
from hrms.payroll.utils import sanitize_expression


class BCNSalaryStructure(SalaryStructure):
    def sanitize_condition_and_formula_fields(self):
        for table in ("earnings", "deductions", "custom_bcn_contributions"):
            for row in self.get(table):
                row.condition = row.condition.strip() if row.condition else ""
                row.formula = row.formula.strip() if row.formula else ""
                row._condition, row.condition = row.condition, sanitize_expression(
                    row.condition
                )
                row._formula, row.formula = row.formula, sanitize_expression(
                    row.formula
                )

    def set_missing_values(self):
        overwritten_fields = [
            "depends_on_payment_days",
            "variable_based_on_taxable_salary",
            "is_tax_applicable",
            "is_flexible_benefit",
        ]
        overwritten_fields_if_missing = ["amount_based_on_formula", "formula", "amount"]
        for table in ["earnings", "deductions", "custom_bcn_contributions"]:
            for d in self.get(table):
                component_default_value = frappe.db.get_value(
                    "Salary Component",
                    cstr(d.salary_component),
                    overwritten_fields + overwritten_fields_if_missing,
                    as_dict=1,
                )
                if component_default_value:
                    for fieldname in overwritten_fields:
                        value = component_default_value.get(fieldname)
                        if d.get(fieldname) != value:
                            d.set(fieldname, value)

                    if not (d.get("amount") or d.get("formula")):
                        for fieldname in overwritten_fields_if_missing:
                            d.set(fieldname, component_default_value.get(fieldname))

    def validate_payment_days_based_dependent_component(self):
        abbreviations = self.get_component_abbreviations()
        for component_type in ("earnings", "deductions", "custom_bcn_contributions"):
            for row in self.get(component_type):
                if (
                    row.formula
                    and row.depends_on_payment_days
                    # check if the formula contains any of the payment days components
                    and any(
                        re.search(r"\b" + abbr + r"\b", row.formula)
                        for abbr in abbreviations
                    )
                ):
                    message = _(
                        "Row #{0}: The {1} Component has the options {2} and {3} enabled."
                    ).format(
                        row.idx,
                        frappe.bold(row.salary_component),
                        frappe.bold(_("Amount based on formula")),
                        frappe.bold(_("Depends On Payment Days")),
                    )
                    message += "<br><br>" + _(
                        "Disable {0} for the {1} component, to prevent the amount from being deducted twice, as its formula already uses a payment-days-based component."
                    ).format(
                        frappe.bold(_("Depends On Payment Days")),
                        frappe.bold(row.salary_component),
                    )
                    frappe.throw(message, title=_("Payment Days Dependency"))

    def validate_formula_setup(self):
        for table in ["earnings", "deductions", "custom_bcn_contributions"]:
            for row in self.get(table):
                if not row.amount_based_on_formula and row.formula:
                    frappe.msgprint(
                        _(
                            "{0} Row #{1}: Formula is set but {2} is disabled for the Salary Component {3}."
                        ).format(
                            table.capitalize(),
                            row.idx,
                            frappe.bold(_("Amount Based on Formula")),
                            frappe.bold(row.salary_component),
                        ),
                        title=_("Warning"),
                        indicator="orange",
                    )

    def reset_condition_and_formula_fields(self):
        # set old values (allowing multiline strings for better readability in the doctype form)
        for table in ("earnings", "deductions", "custom_bcn_contributions"):
            for row in self.get(table):
                row.condition = row._condition
                row.formula = row._formula

        self.db_update_all()
